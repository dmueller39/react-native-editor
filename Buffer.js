// @flow
import React, { PureComponent } from 'react';
import { FlatList } from 'react-native';

import {
  type Line,
  type Lines,
  type RawLocation,
  isLocationEqual,
} from './types';

import EditableLine from './EditableLine';
import BufferLine from './BufferLine';

import { processLines, getLineLayout } from './util';

const defaultProps = {
  selectedWord: null,
  selectedLineIndex: null,
  activeLineText: null,
  data: '',
  width: 0,
  selectedLocation: null,
};

type DefaultProps = typeof defaultProps;

type Props = {
  onSelectLine: (number) => void,
  onSelectWord: (string, RawLocation) => void,
  onSelectSelectedWord: () => void,
  onChangeActiveLine: (string) => void,
  onDeleteNewline: () => void,
  data: string,
  selectedWord: ?string,
  selectedLineIndex: ?number,
  activeLineText: ?string,
  width: number,
  selectedLocation: ?RawLocation,
};

export default class Buffer extends PureComponent<DefaultProps, Props, void> {
  static defaultProps = defaultProps;

  constructor(props: Props) {
    super();
    this.lines = processLines(
      props.data,
      props.width,
      props.selectedWord,
      props.selectedLocation,
      props.selectedLineIndex,
      props.activeLineText
    );
  }

  componentWillReceiveProps(nextProps: Props) {
    this.lines = processLines(
      nextProps.data,
      nextProps.width,
      nextProps.selectedWord,
      nextProps.selectedLocation,
      nextProps.selectedLineIndex,
      nextProps.activeLineText
    );
  }
  componentDidUpdate(prevProps: Props) {
    if (
      this.props.selectedLocation != null &&
      prevProps.selectedLocation != null &&
      isLocationEqual(this.props.selectedLocation, prevProps.selectedLocation)
    ) {
      this.scrollToRawLineIndex(this.props.selectedLocation.lineIndex);
    }

    if (
      this.props.selectedLineIndex != null &&
      this.props.selectedLineIndex !== prevProps.selectedLineIndex
    ) {
      this.scrollToRawLineIndex(this.props.selectedLineIndex);
    }
  }

  onViewableItemsChanged = (info: { viewableItems: Array<{ item: Line }> }) => {
    this.viewableItems = info.viewableItems;
  };

  flatListRef: FlatList<Line>;
  lines: Lines = [];
  viewableItems: Array<{ item: Line }> = [];
  scrollToRawLineIndex(rawLineIndex: number) {
    const viewableSelectedItem = this.viewableItems.find(
      viewableItem => viewableItem.item.rawLineIndex === rawLineIndex
    );

    // don't scroll if the line is already visible
    if (viewableSelectedItem == null) {
      const index = this.lines.findIndex(
        line => line.rawLineIndex === rawLineIndex
      );
      this.flatListRef.scrollToIndex({
        index,
        viewPosition: 0.5,
        animated: false,
      });
    }
  }

  captureRef = (ref: FlatList<Line>) => {
    this.flatListRef = ref;
  };

  renderItem = (param: { item: Line, index: number }) => {
    if (param.item.isEditing && param.item.activeLineText != null) {
      return (
        <EditableLine
          onChangeActiveLine={this.props.onChangeActiveLine}
          onDeleteNewline={this.props.onDeleteNewline}
          text={param.item.activeLineText}
        />
      );
    }
    return (
      <BufferLine
        selectedLineIndex={this.props.selectedLineIndex}
        onSelectLine={this.props.onSelectLine}
        onSelectWord={this.props.onSelectWord}
        onSelectSelectedWord={this.props.onSelectSelectedWord}
        {...param}
        lines={this.lines}
      />
    );
  };

  render() {
    return (
      <FlatList
        {...this.props}
        ref={this.captureRef}
        getItemLayout={getLineLayout}
        onViewableItemsChanged={this.onViewableItemsChanged}
        showsVerticalScrollIndicator={false}
        keyExtractor={(item: void, index: number) => `line-${String(index)}`}
        windowSize={5}
        data={this.lines}
        renderItem={this.renderItem}
      />
    );
  }
}
