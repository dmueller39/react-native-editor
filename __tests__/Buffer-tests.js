// @flow
import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

import Buffer from '../Buffer';

it('renders correctly', () => {
  const tree = renderer.create(
    <Buffer
      onSelectLine={() => {}}
      onSelectWord={() => {}}
      onChangeData={() => {}}
      data="foo"
      selectedWord={null}
      selectedLineIndex={null}
      width={320}
      selectedLocation={null}
      isEditing={false}
    />
  );

  expect(tree).toMatchSnapshot();
});
