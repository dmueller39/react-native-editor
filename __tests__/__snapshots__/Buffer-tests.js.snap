// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  activeLineText={null}
  data={
    Array [
      Object {
        "activeLineText": null,
        "continued": false,
        "continuing": false,
        "end": 3,
        "rawLineIndex": 0,
        "start": 0,
        "text": "foo",
        "textSections": Array [
          Object {
            "end": 3,
            "highlight": "normal",
            "start": 0,
            "text": "foo",
          },
        ],
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  getItemLayout={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onChangeActiveLine={[Function]}
  onContentSizeChange={[Function]}
  onDeleteNewline={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onSelectLine={[Function]}
  onSelectSelectedWord={[Function]}
  onSelectWord={[Function]}
  onViewableItemsChanged={[Function]}
  renderItem={[Function]}
  renderScrollComponent={[Function]}
  scrollEventThrottle={50}
  selectedLineIndex={null}
  selectedLocation={null}
  selectedWord={null}
  showsVerticalScrollIndicator={false}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  width={320}
  windowSize={5}
>
  <View>
    <View
      onLayout={undefined}
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        hitSlop={undefined}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={undefined}
        testID={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {},
              Object {
                "fontFamily": "Courier New",
                "fontSize": 14,
              },
            ]
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            disabled={false}
            ellipsizeMode="tail"
            style={Object {}}
          >
            foo
          </Text>
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
